### ISSUE TITLE
[idea] let the library wrap the SNTP SDK functions of ESP8266 and ESP32

### ISSUE THREAD

**Issue by JAndrassy** [gmag11/NtpClient#112]
I think this library is still useful. It could instead of retriving the time with UDP on esp8266 and esp32 work with the SDK API. The API is different on the esps so this way this library would hide the differences and have one way for different hardware.
And additionally to retrieving the time from NTP server this library sets-up the TimeLib. It helps the IoT newbies much.

Some thoughts to compatibility with different networking libraries. I made my local copy of this library compatible with WiFiNINA and my WiFiEspAT library simply by adding NETWORK_WIFI. (My [PR to add WiFi.h to WiFi101](https://github.com/arduino-libraries/WiFi101/pull/291) was not accepted yet so NETWORK_WIFI101 is still necessary)
```
#elif NETWORK_TYPE == NETWORK_WIFI
#include <WiFi.h>
#include <WiFiUdp.h>
```

a tip to network library used by the sketch detection. `__has_include` works
```
#if __has_include(<Ethernet.h>)
#include <Ethernet.h>
#define USE_ETHERNET 1
#else
#include <WiFi.h>
#endif
```

In my TelnetStream library [examples](https://github.com/jandrassy/TelnetStream/tree/master/examples) I played with esp SDK time retrieving. it is simple, but not so straightforward as with NtpClient library.

EDIT:
I see this in NtpClientLib.h 
```
#elif NETWORK_TYPE == NETWORK_WIFI101
#include <WiFiClient.h>
#include <WiFiUdp.h>
#include <WiFi101.h>
```
always include the specific include first (WiFi101.h). Used in this ordering in a sketch it would include WiFiClient of the old WiFi library which is still bundled with Arduino IDE.

**Comment by gmag11** [comment]
I've been testing internal SNMP implementation on ESP32 and ESP8266. They are similar to current NTPClient implementation in terms of precision.
For a current project I need at least millisecond sync, so I will go on with current implementation and improve NTP packet decoding to get time fraction field.
Thank you for your comments. I'll keep them on mind.

Are you still using this library for something different than Espressif microcontrollers?

**Comment by JAndrassy** [comment]
I did use this library in examples of [my TelnetStream library](https://github.com/jandrassy/TelnetStream) to log with timestams. But I removed it because there where problems with architecture/networking pairs like AVR/Ethernet, MKR/Wifi while people use AVR with WiFi and MKR with Ethernet too.

I have [this header file](https://github.com/jandrassy/TelnetStream/blob/master/src/NetTypes.h) to detect the networking library used by the sketch. It is not perfect because esp platforms  don't have '__has_include', but it covers many combinations.

I read in README of this library that the esp RTC time is used to evaluate 'SSL' certificates so it should be set. But at least for esp8266 it doesn't look like there is a way to set RTC time other then to configure SNTP.

**Comment by gmag11** [comment]
Yes, all TLS related code using core fuctions make use of internal SDK time implementation to check certificate validity. So if another mechanism (like TimeLib.h) to track calendar time TLS will fail to connect.
My idea is to rewrite this library from scratch to use internals for time keeping. I've found several problems in AsyncUDP library in ESP32 so I will change library architecture quite a bit.

I find very difficult to keep code working on all platforms. I could do automated testing to find bugs but I do not have enough skills on that matter. So, I decided to move next major version (4.0.0) to ESP8266 and ESP32 only.

Whan I am implementing would be easily migrated to other platforms so I encourage anybody to take my code an give it another use.

New version will include a much more complete NTP offset calculation so it can have near 1ms time sync precision.
